geom/twkb_write.go

```diff
@@ -0,0 +1,535 @@
+	}
+	return nil
+}
+
+func (w *TWKBWriter) writeGeometryCollection(gc GeometryCollection) error {
+	w.writeTypeAndPrecision(twkbTypeGeometryCollection)
+
+	if ctype := gc.CoordinatesType(); ctype != w.ctype {
+		return fmt.Errorf("mismatched coordinate dimensions got %s expected %s", ctype, w.ctype)
+	}
+
+	if gc.IsEmpty() {
+		w.writeIsEmptyHeader()
+		return nil
+	}
+	w.writeInitialHeaders()
+
+	numGeometries := gc.NumGeometries()
+	w.writeUnsignedVarint(uint64(numGeometries))
+
+	if err := w.writeIDList(numGeometries); err != nil {
+		return err
+	}
+
+	for i := 0; i < numGeometries; i++ {
+		subWriter := copyTWKBWriter(w)
+		geom := gc.GeometryN(i)
+		subWriter.writeGeometry(geom)
+		w.twkb = append(w.twkb, subWriter.twkb...)
+	}
+	return nil
+}
+
+func (w *TWKBWriter) writeTypeAndPrecision(kind twkbGeometryType) {
+	w.kind = kind
+	w.writeByte(byte(EncodeZigZagInt32(int32(w.precXY))<<4) | byte(w.kind))
+}
+
+func (w *TWKBWriter) writeIsEmptyHeader() {
+	w.isEmpty = true
+	w.writeMetadataHeader(twkbIsEmpty)
+	// Do not write any extended info, bbox, size, or ids.
+}
+
+func (w *TWKBWriter) writeInitialHeaders() {
+	var metaheader twkbMetadataHeader
+	if w.hasExt {
+		metaheader |= twkbHasExtPrec
+	}
+	if w.hasSize {
+		metaheader |= twkbHasSize
+	}
+	if w.hasBBox {
+		metaheader |= twkbHasBBox
+	}
+	if w.hasIDs {
+		metaheader |= twkbHasIDs
+	}
+	w.writeMetadataHeader(metaheader)
+
+	if w.hasExt {
+		w.writeExtendedPrecision()
+	}
+}
+
+func (w *TWKBWriter) writeMetadataHeader(metaheader twkbMetadataHeader) {
+	if metaheader&twkbIsEmpty != 0 {
+		w.isEmpty = true
+	}
+	w.writeByte(byte(metaheader))
+}
+
+func (w *TWKBWriter) writeExtendedPrecision() {
+	var ext byte
+	if w.hasZ {
+		ext |= 0x01
+		ext |= byte(EncodeZigZagInt32(int32(w.precZ)) << 2)
+	}
+	if w.hasM {
+		ext |= 0x02
+		ext |= byte(EncodeZigZagInt32(int32(w.precM)) << 5)
+	}
+	w.writeByte(ext)
+}
+
+func (w *TWKBWriter) writePoints(numPoints int, coords ...float64) {
+	w.writePointArray(numPoints, coords)
+}
+
+// Convert a given number of points from floating point to integer coordinates.
+// Utilise and update the running memory of the previous reference point.
+// The input coords must contain numPoints * the number of dimensions values.
+func (w *TWKBWriter) writePointArray(numPoints int, coords []float64) {
+	var buf [binary.MaxVarintLen64]byte
+	c := 0
+	for i := 0; i < numPoints; i++ {
+		for d := 0; d < w.dimensions; d++ {
+			fval := coords[c]
+			ival := int64(fval * w.scalings[d])
+			// Compute bounding box.
+			if !w.bboxValid {
+				w.bboxMin[d] = ival
+				w.bboxMax[d] = ival
+			} else if ival < w.bboxMin[d] {
+				w.bboxMin[d] = ival
+			} else if ival > w.bboxMax[d] {
+				w.bboxMax[d] = ival
+			}
+			// Perform coord differencing to find the int value.
+			ival -= w.refpoint[d]
+			n := binary.PutVarint(buf[:], ival)
+
+			w.twkb = append(w.twkb, buf[:n]...)
+			w.refpoint[d] += ival
+			c++
+		}
+		if !w.bboxValid {
+			w.bboxValid = true
+		}
```

Comment:
Maybe I'm missing something but why is there a need to check `if !w.bboxValid`?

Reply:
Yep, you're right, that was a remnant of some refactored code. It was originally inside the inner loop within an `if` statement having the same condition, but that was incorrect because it would set the `bboxValid` flag to true after the X co-ordinate was examined, rather than after the entire first point was examined (which had a weird side-effect of making the bounding box min and max extents wrong for all dimensions except for X).

I'll add a comment to explain why the bbox becomes valid after the first point, i.e. why the code is where it is.